### ISSUE TITLE
[no-non-null-asserted-optional-chain] Possible Infinite Loop

### ISSUE THREAD

**Issue by hjkcai** [typescript-eslint/typescript-eslint#2521]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-non-null-asserted-optional-chain": "error"
  }
}
```

```TS
let foo1: { bar: { baz: string } } | undefined;

console.log(foo1?.bar!.baz);
```

**Expected Result**

ESLint reports error.

**Actual Result**

Infinite Loop.

**Additional Info**

I am using ESLint programmatically:

```ts
import { CLIEngine, Linter } from 'eslint';
const cli = new CLIEngine({});
cli.executeOnFiles('/path/to/the/file.ts');
```

But the script never ended. However this problem was not present in VSCode ESLint Plugin. I have no idea about what's happening. I am not sure the above code can reproduce the problem (because it is extracted from a large codebase). 

I used the debugger to pause when infinite loop occurred. It ended up here:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-non-null-asserted-optional-chain.ts#L116

There is an obvious logical error: the `break` only breaks the `switch` statement, meanwhile `current.optional` is truthy, meaning `current` is truthy. Theoretically **THE LOOP WILL NEVER STOP WHEN A OPTIONAL CHAIN IS FOUND**.

```ts
while (current) {
  switch (current.type) {
    case AST_NODE_TYPES.MemberExpression:
      if (current.optional) {
        // found an optional chain! stop traversing
        break;
      }

      current = current.object;
      continue;

    case AST_NODE_TYPES.CallExpression:
      if (current.optional) {
        // found an optional chain! stop traversing
        break;
      }

      current = current.callee;
      continue;

    default:
      // something that's not a ChainElement, which means this is not an optional chain we want to check
      return;
  }
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.18.2` |

**Comment by bradzacher** [comment]
We have this exact code as a passing test case, so I'm not sure why you're encountering an infinite loop.

https://github.com/typescript-eslint/typescript-eslint/blob/762bc99584ede4d0b8099a743991e957aec86aa8/packages/eslint-plugin/tests/rules/no-non-null-asserted-optional-chain.test.ts#L21

The only reason I could see is because you're on an old version of TS, which triggers that code path.

Though I'm not sure why nobody else has encountered this?

Could you provide the full config you're using?

**Comment by hjkcai** [comment]
Thanks for reply. I will try to make a minimal reproduction. 

BTW, the possible infinite loop is still exist in the code path. Maybe we need to figure out what triggers it.

**Comment by bradzacher** [comment]
Our tests only run on the latest TS version because nobody's spent the time to investigate / build out infra for multi-version testing.

Hence the passing test.

You're running TS 3.7, which uses a different code-path to the latest TS version (they made a breaking change to non-null asserted optional chains in TS 3.9).

Though yeah I'm not sure why nobody else as run into an infinite loop.

**Comment by hjkcai** [comment]
Yes. TS 3.7 causes the problem. I opened an individual project using:

```json
{
  "dependencies": {
    "@typescript-eslint/eslint-plugin": "^4.1.0",
    "@typescript-eslint/parser": "^4.1.0",
    "eslint": "^6.8.0",
    "typescript": "^3.7.5"
  }
}
```

```json
{
  "compilerOptions": {
    "target": "es2018",
    "strict": true
  }
}
```

```json
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json'
  },
  plugins: ['@typescript-eslint'],
  rules: {
    "@typescript-eslint/no-non-null-asserted-optional-chain": "error"
  }
}
```

```ts
// test.ts
let foo1: { bar: { baz: string } } | undefined;

console.log(foo1?.bar!.baz);
foo1?.bar!.baz
```

Executing eslint never stops.

```bash
npx eslint test.ts
```

I will upgrade my project to a higher version of TypeScript firstly. Take time to solve this problem 😄

**Comment by tadhgmister** [comment]
It may be desirable to add an internal  `forceUsing3.7Logic` option so testing can hit the old logic. currently that [whole section is untested](https://codecov.io/gh/typescript-eslint/typescript-eslint/pull/2380/diff?src=pr&el=tree#D12-100).

**Comment by bradzacher** [comment]
I'd rather implement a full solution for testing old TS versions rather than adding per-rule hacks.
See discussion in #1752

**Comment by JoshuaKGoldberg** [comment]
#5915 is merged into #5886; closing this issue now accordingly. It'll be released as a part of the v6 major release.
